[RISCV] Fold PseudoVMV_V_V with undef passthru, handling policy

lukel97 opened the pull request:
If a vmv.v.v has an undef passthru then we can just replace it with its input operand, since the tail is completely undefined.

This is a reattempt of #106840, but also checks to see if the input was a pseudo where we can relax its tail policy to undef.

This also means we don't need to check for undef passthrus in foldVMV_V_V anymore because they will be handled by foldUndefPassthruVMV_V_V.

preames commented on llvm/test/CodeGen/RISCV/rvv/vmv.v.v-peephole.ll:
```diff
@@ -194,3 +194,14 @@ define <vscale x 2 x i32> @unfoldable_mismatched_sew(<vscale x 2 x i32> %passthr
+; CHECK-NEXT:    vsetvli zero, a0, e64, m1, tu, ma
+; CHECK-NEXT:    vadd.vv v8, v9, v10
+; CHECK-NEXT:    ret
+  %a = call <vscale x 1 x i64> @llvm.riscv.vadd.nxv1i64.nxv1i64(<vscale x 1 x i64> %passthru, <vscale x 1 x i64> %x, <vscale x 1 x i64> %y, iXLen %avl)
```
Does this actually hold?  Consider:
SrcVL = 2, so elements 0, 1 come from either x or y, but 2 onwards come from passthru
MIVL = 4, so elements 0, 1 come from either x or y, elements 2, 3 come from passthru, and elements 4+ are undefined

Oh, I see it.  You're not changing passthru or VL on Src, you're basically just dropping MI.

I think this might be cleaner if expressed as an early continue here.  This is arguably a different combine entirely.  You're basically just using properties of the vmv.v.v itself to prove it can be elided entirely.  Actually, now that I say that, yeah, this really doesn't depend on Src at all does it?  I don't think you even need the single use condition.

lukel97 replied:
Yup, that's exactly what I did in #106840. I added a new peephole that doesn't check Src, but I closed it because I thought it was too similar to foldVMV_V_V.

However I've since realised that if the vmv.v.v passthru is undef, then we can still relax Src's policy to be tail agnostic if vmv.v.s tail subsumes it. E.g. if we have

```
vsetvli zero, a0, e64, m1, tu, ma
vadd.vv v8, v9, v10
vsetvli zero, zero, e64, m1, ta, ma
vmv.v.v v8, v8
```

We should actually be able to fold it into:

```
vsetvli zero, a0, e64, m1, ta, ma
vadd.vv v8, v9, v10
```

Hence why I moved it back into foldVMV_V_V, so it could take advantage of #105788.

But I completely agree with you here that this is could be much cleaner. I'm going to try again to do this as a separate peephole, but explicitly relaxing Src's policy if possible.

preames reviewed (commented):
Looks functionally correct, but would be much more cleanly expressed as an early return.  This is really just a different transform entirely.

lukel97 committed:
[RISCV] Fold PseudoVMV_V_V with undef passthru, handling policy

If a vmv.v.v has an undef passthru then we can just replace it with its input operand, since the tail is completely undefined.

This is a reattempt of #106840, but also checks to see if the input was a pseudo where we can relax its tail policy to undef.

This also means we don't need to check for undef passthrus in foldVMV_V_V anymore because they will be handled by foldUndefPassthruVMV_V_V.

lukel97 commented:
I've reworked this PR and updated the description to do this as a separate peephole, as I think it's easier to understand without having to think about foldVMV_V_V.

preames commented on llvm/lib/Target/RISCV/RISCVVectorPeephole.cpp:
```diff
@@ -472,6 +473,38 @@ bool RISCVVectorPeephole::ensureDominates(const MachineOperand &MO,
+      SrcPolicy.setImm(SrcPolicy.getImm() | RISCVII::TAIL_AGNOSTIC);
+  }
+
+  MRI->replaceRegWith(MI.getOperand(0).getReg(), Src->getOperand(0).getReg());
```
Src may have multiple defs, use MI.getOperand(2).getReg() instead.

preames commented on llvm/test/CodeGen/RISCV/rvv/vmv.v.v-peephole.ll:
```diff
@@ -194,3 +194,14 @@ define <vscale x 2 x i32> @unfoldable_mismatched_sew(<vscale x 2 x i32> %passthr
   ret <vscale x 2 x i32> %b
 }
+
+define <vscale x 1 x i64> @undef_passthru(<vscale x 1 x i64> %passthru, <vscale x 1 x i64> %x, <vscale x 1 x i64> %y, iXLen %avl) {
```
JFYI, I precommitted these tests, so make sure you rebase.

(I'd had another idea for an approach here, but after writing it decided this was fine, and I'll do a followup to generalize and simplify.)

preames reviewed (approved):
LGTM w/minor changes.

lukel97 committed:
Address comments, also don't attempt further folds if MI was deleted
